packages/SwingSet/src/liveslots/collectionManager.js

```diff
@@ -200,6 +201,24 @@ export function makeCollectionManager(
     const dbKeyPrefix = `vc.${collectionID}.`;
     let currentGenerationNumber = 0;
 
+    const keyLabel = `invalid key type for collection ${q(label)}`;
+    const valueLabel = `invalid value type for collection ${q(label)}`;
+
+    const serializeValue = value => {
+      if (valueShape !== undefined) {
+        fit(value, valueShape, valueLabel);
+      }
+      return serialize(value);
+    };
+
+    const unserializeValue = data => {
+      const value = unserialize(data);
+      if (valueShape !== undefined) {
+        fit(value, valueShape, valueLabel);
+      }
```

Comment:
Given that we check on store, we probably should not redundantly check on read.

Is there any similar issue when reading data stored by a previous incarnation?

Reply:
I think data from a previous incarnation could be a thing (e.g., an upgrade adds a pattern that wasn't previously there), so I'd be inclined to check on both read and write.  Our general policy in the kernel/liveslots world is to be paranoid and check stuff all over the place, unless and until performance measurement tells it that it's costing too much.

Reply:
Good to know. I will continue to check on read until we have evidence of a problem.

Reply:
Resolving.